raw-ipa-lib/src/ss.rs

```diff
@@ -0,0 +1,211 @@
+use curve25519_dalek_ng::constants;
+use rand::thread_rng;
+use rust_elgamal::{Ciphertext, DecryptionKey, EncryptionKey, RistrettoPoint, Scalar};
+
+#[derive(Clone)]
+pub struct GeneratorShare {
+    p: RistrettoPoint,
+}
+
+#[derive(Clone)]
+pub struct EncryptedGeneratorShare {
+    c: Ciphertext,
+}
+
+impl GeneratorShare {
+    #[must_use]
+    pub fn share(value: u8) -> (Self, Self) {
+        let mut rng = thread_rng();
+        let v = Scalar::from(value);
+        let mapped_value = &v * &constants::RISTRETTO_BASEPOINT_TABLE;
+        let ss1 = RistrettoPoint::random(&mut rng);
+        let ss2 = mapped_value - ss1;
+        (Self { p: ss1 }, Self { p: ss2 })
+    }
+
+    #[must_use]
+    pub fn encrypt(&self, ek: &EncryptionKey) -> EncryptedGeneratorShare {
+        let mut rng = thread_rng();
+
+        let r = Scalar::random(&mut rng);
+        EncryptedGeneratorShare {
+            c: ek.encrypt_with(self.p, r),
+        }
+    }
+
+    #[must_use]
+    pub fn pt(&self) -> RistrettoPoint {
+        self.p
+    }
+
+    /// # Panics
+    /// If this point does not correspond to a value between 1 and 1000 (inclusive)
+    #[must_use]
+    pub fn map_point_to_value(r: &RistrettoPoint) -> u32 {
+        let mut i = 1;
+        while i <= 1000 {
+            let query = &Scalar::from(i) * &constants::RISTRETTO_BASEPOINT_TABLE;
+            if query.eq(r) {
+                return i;
+            }
+            i += 1;
+        }
+        panic!("point does not map to a value between 1 and 1000");
```

Comment:
That would definitely be better =). This was pretty much just test code to confirm if this approach even worked =). Now that I see that it can - I'll clean up this lame code =).